Refactor mailer previews

megcrenshaw opened the pull request:
This is a pre-cursor to https://codedotorg.atlassian.net/browse/PLAT-1848, which is to update AYW content in mailers.

The methods in `mailers/previews/pd_workshop_mailer_preview.rb` were getting more and more challenging to keep track of, and some methods were producing mailers for workshops that we no longer have.

I removed all ECS, CS in A, and CS in S mailers based on the following:
```
[production] dashboard > Pd::Workshop.where(course: 'Exploring Computer Science', ended_at: nil).pluck(:id)
=> [6304, 6305, 6306, 8632]

[production] dashboard > Pd::Workshop.where(course: 'CS in Science', ended_at: nil).pluck(:id)
=> []

[production] dashboard > Pd::Workshop.where(course: 'CS in Algebra', ended_at: nil).pluck(:id)
=> []
```
— all of the ECS were created in 2015 (ids 6304, 6305, 6306) or 2020 (id 8632). The one in 2020 (https://studio.code.org/pd/workshop_dashboard/workshops/8632) already had their workshop

Additionally, some of the namespace work that @Hamms did changed the URLs for the mailers––I updated those to reflect that change.

## Links

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story

https://staging-studio.code.org/rails/mailers/pd_workshop_mailer shows 80 emails. From there, I ... 
- Removed 9 receipts:
    - 6 Exploring Computer Science receipts (one called “phase_2”)
    - 2 CS in Science receipts
    - 1 CS in Algebra receipt
- Removed 1 “csd_unit_6_3_day” reminder (duplicate with other AYW reminders)
- Added 2 CSF Virtual receipts (deepdive and intro)
- Added 2 CSP virtual AYW reminders (3 day and 10 day)
- Added 2 CSD virtual AYW reminders (3 day and 10 day)
- Added 6 CSA workshops (virtual and not, receipt + 3 day reminder + 6 day reminder)

That's a net of 2 added emails, so 82 emails total. There are 82 emails on my local machine (http://localhost-studio.code.org:3000/rails/mailers/pd_workshop_mailer). Additionally, I spot-checked some mailers to ensure everything looked good.

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

davidsbailey commented:
> — all of the ECS were created in 2015 (ids 6304, 6305, 6306) or 2020 (id 8632). The one in 2020 (https://studio.code.org/pd/workshop_dashboard/workshops/8632) already had their workshop

maybe obvious but just confirming, already had all their workshop emails sent?

megcrenshaw commented:
@davidsbailey I didn't check that, but I can ... two things to note here:
(1) these previews don't affect what actually can get mailed; they just populate the list at `/rails/mailers/pd_workshop_mailer` and those get tested
(2) Let's say their workshop emails were *not* sent. Would we be sending them 2 years after they were created? If yes, we could still send them––we'd just need to do a bit more work to get the previews shown if someone wants to see that before sending.

davidsbailey commented:
hey Meg, apologies for my brevity, I was just having trouble understanding the text "already had their workshop" since it seemed there might be more to the sentence. I get what you are saying now that the workshop already took place. no need to check that emails were sent, as far as I'm concerned -- agree that if they were not sent, we would not go back and send them now.

megcrenshaw commented:
Sounds good. I was mostly just ensuring that nobody will be looking for these previews, but yeah that sentence was confusing!

megcrenshaw commented on bin/oneoff/diff_email_previews:
```diff
@@ -16,21 +16,22 @@ To diff emails in 2 folders in the current directory: './diff_email_previews dif
 # List of email previews are copied from
-# http://localhost-studio.code.org:3000/rails/mailers/pd/workshop_mailer/
+# http://localhost-studio.code.org:3000/rails/mailers/pd_workshop_mailer/
 PREVIEWS = %w(
```
This whole list I re-copied and pasted from my localhost. I'm not sure how often we use this, but figured I'd update it

dmcavoy commented on lib/cdo/shared_constants/pd/shared_workshop_constants.rb:
```diff
@@ -27,7 +27,7 @@ module SharedWorkshopConstants
     SUBJECT_TEACHER_CON = 'Code.org TeacherCon'.freeze
 
-    # Academic Year Workshop subjects shared between CSP and CSD.
+    # Academic Year Workshop subjects shared between CSP, CSD, and CSP
```
Should this say CSA, CSD and CSP ?

megcrenshaw replied:
Yikes! Good catch, thanks

dmcavoy commented on dashboard/test/mailers/previews/pd_workshop_mailer_preview.rb:
```diff
@@ -1,90 +1,53 @@
-        location_address: nil
-      }
-  end
+      define_method("teacher_enrollment_receipt__#{partial}_virtual") do
```
Nice clean up here!

dmcavoy commented on dashboard/app/mailers/pd/workshop_mailer.rb:
```diff
@@ -10,22 +10,21 @@ class Pd::WorkshopMailer < ActionMailer::Base
-      Pd::Workshop::SUBJECT_CS_IN_S_PHASE_3_SEMESTER_2 => 'cs_in_s_phase_3_semester_2'
+    Pd::Workshop::COURSE_CSF => {
+      Pd::Workshop::SUBJECT_CSF_101 => 'csf_intro',
+      Pd::Workshop::SUBJECT_CSF_201 => 'csf_deepdive'
```
This is making me wonder if we need something here for the district workshops. I did not get a chance to check the emails for that.

dmcavoy replied:
@megcrenshaw any thoughts on this based on what you have looked at here?

megcrenshaw replied:
Hmm, nothing in https://staging-studio.code.org/rails/mailers/pd_workshop_mailer has the district workshops.

I did a quick Jira search and couldn't find anything about mailers for district workshops (https://codedotorg.atlassian.net/browse/PLAT-1836 added them)

megcrenshaw replied:
AH, you asked this question earlier, and it's the same as other CSF subjects (verified with Tess in a DM)! I'll add this one subject as well to make that clear.

megcrenshaw replied:
(May 26 was when this was confirmed)

dmcavoy replied:
Let me check in with Danielle on what she wants for emails. The post workshop email is working because I was able to test that. Not sure about the 3 and 10 day. Thanks!

dmcavoy commented on bin/oneoff/diff_email_previews:
```diff
@@ -16,21 +16,22 @@ To diff emails in 2 folders in the current directory: './diff_email_previews dif
   detail_change_notification__admin
   detail_change_notification__csf_deepdive
   detail_change_notification__csf_intro
-  exit_survey__csd_1
```
I'm wondering about exit_survey__csd_1 and exit_survey__csp_1 which have been deleted here

megcrenshaw replied:
The list here (in diff_email_previews) was last updated in 2020. The list we currently have is at https://staging-studio.code.org/rails/mailers/pd_workshop_mailer, which does *not* have CSD or CSP exit surveys

dmcavoy replied:
I'm wondering if those emails had to do with TeacherCon since they are next to it in the list. Be interested to know what the email associated with it said

megcrenshaw replied:
I'm wondering if this oneoff script should even be updated ... I can restore this file if you think it's best to leave it untouched (no mentions of it in Slack)

dmcavoy replied:
Seems like if it has not been on staging for a while it's probably fine to go ahead without it.

dmcavoy reviewed (approved):
Changes look good to me. Interested to hear your thoughts on the District workshop emails based on what you have looked at with this PR

megcrenshaw commented:
Thanks for looking, Dani! Just fixed the typo in the comment and added the district mailer.

I think if we do need to update the content in the district mailer, that should be part of a separate PR that goes with updating content for AYWs.
